### ISSUE TITLE
Corrupted strings on save + reload with multi-byte UTF-8 characters with ActionText::RichText on Rails 7.1

### ISSUE THREAD

**Issue by kaoru** [ankane/lockbox#183]
We're having a problem when upgrading to Rails 7.1 and using lockbox to encrypt ActionText rich text content fields.

I'm seeing multi-byte UTF-8 characters getting corrupted on roundtrip encrypt/save/load/decrypt.

I've written a failing test and a possible fix here: https://github.com/CharlieHR/lockbox/commit/0fe861e9b90b6fd444aa7baacad98d142c72fa0c 

The test is good:

```
  def test_encoding
    user = User.create!(content: "ルビー")
    assert_equal user.content.to_trix_html, "ルビー"
    assert_equal user.reload.content.to_trix_html, "ルビー"
  end
```

But I'm not very confident in the fix because I don't fully understand what's going on with activemodel type serialization and how that interacts with lockbox.

-----------

Here's some more details about what i've observed as differences:

On Rails 7.0.8, `type.deserialize(message)` correctly produces the correctly encoded string:

```
# Rails 7.0.8

> message
=> "We\xE2\x80\x99re very excited to have you onboard."

> type
=> #<ActiveModel::Type::String:0x000000013e4390b0 @false="f", @limit=nil, @precision=nil, @scale=nil, @true="t">

> type.deserialize(message).to_s
=> "<div class=\"trix-content\">\n  We’re very excited to have you onboard.\n</div>\n"
```

But on Rails 7.1.2 `type.deserialize(message)` corrupts the data:

```
# Rails 7.1.2

> message
=> "We\xE2\x80\x99re very excited to have you onboard."

> type
=> #<ActiveModel::Type::String:0x0000000109c94988 @false="f", @limit=nil, @precision=nil, @scale=nil, @true="t">

> type.deserialize(message).to_s
=> "<div class=\"trix-content\">\n  Weâ\u0080\u0099re very excited to have you onboard.\n</div>\n"
```

I noticed one other very strange thing - the type is an instance of `ActiveRecord::Type::Serialized`, which doesn't return true for the `is_a?(ActiveModel::Type::String)` check even though it does call itself a `ActiveModel::Type::String` when you call `#to_s` on it... something to do with subtypes I think?

```
> type
=> #<ActiveModel::Type::String:0x000000014bef6a40 @false="f", @limit=nil, @precision=nil, @scale=nil, @true="t">

> type.class
=> ActiveRecord::Type::Serialized

> puts type.inspect
#<ActiveRecord::Type::Serialized:0x000000014e818978 @subtype=#<ActiveModel::Type::String:0x000000014bef6a40 @true="t", @false="f", @precision=nil, @scale=nil, @limit=nil>, @coder=ActionText::Content, @delegate_dc_obj=#<ActiveModel::Type::String:0x000000014bef6a40 @true="t", @false="f", @precision=nil, @scale=nil, @limit=nil>>

> type.subtype.class
=> ActiveModel::Type::String
```

A fix is to do the `message.force_encoding` _before_ the `type.deserialize`...

```
> message
=> "We\xE2\x80\x99re very excited to have you onboard."

> message.force_encoding(Encoding::UTF_8)
=> "We’re very excited to have you onboard."

> type.deserialize(message).to_s
=> "<div class=\"trix-content\">\n  We’re very excited to have you onboard.\n</div>\n"
```

Which is what I tried in my commit: https://github.com/CharlieHR/lockbox/commit/0fe861e9b90b6fd444aa7baacad98d142c72fa0c

**Comment by ankane** [comment]
Hi @kaoru, thanks for reporting and investigating! It looks like the test passes by adding `|| type.try(:subtype).is_a?(ActiveModel::Type::String)` to the current line (without changing the order). Does that fix it for you?

**Comment by kaoru** [comment]
Hey @ankane,

When I run it with just the `|| type.try(:subtype).is_a?(ActiveModel::Type::String)` I get a NoMethodError against the lockbox test suite and our app's full test suite.

```
$ ruby /Users/alex/Documents/Git/stuff/lockbox/test/action_text_test.rb 
Run options: --seed 20162

# Running:

.E

Error:
ActionTextTest#test_encoding:
NoMethodError: undefined method `force_encoding' for #<ActionText::Content "ã\u0083«ã\u0083\u0093ã\u0083¼">
    /Users/alex/Documents/Git/stuff/lockbox/lib/lockbox/model.rb:619:in `block (3 levels) in has_encrypted'
    /Users/alex/Documents/Git/stuff/lockbox/lib/lockbox/model.rb:481:in `block (3 levels) in has_encrypted'
    /Users/alex/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/actiontext-7.1.2/app/models/action_text/rich_text.rb:24:in `nil?'
    /Users/alex/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/activerecord-7.1.2/lib/active_record/associations/association.rb:70:in `reload'
    /Users/alex/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/activerecord-7.1.2/lib/active_record/associations/singular_association.rb:11:in `reader'
    /Users/alex/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/activerecord-7.1.2/lib/active_record/associations/builder/association.rb:104:in `rich_text_content'
    /Users/alex/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/actiontext-7.1.2/lib/action_text/attribute.rb:40:in `content'
    /Users/alex/Documents/Git/stuff/lockbox/test/action_text_test.rb:24:in `test_encoding'

bin/rails test /Users/alex/Documents/Git/stuff/lockbox/test/action_text_test.rb:21

.
```

I think on the latest Rails the call to `type.deserialize(message)` is returning an `ActionText::Content` object that doesn't respond to `force_encoding`.

Could it be that you have an older version of Rails in your Gemfile.lock (which is git ignored)? Here's mine in case that's it:

```
PATH
  remote: .
  specs:
    lockbox (1.3.0)

GEM
  remote: https://rubygems.org/
  specs:
    actioncable (7.1.2)
      actionpack (= 7.1.2)
      activesupport (= 7.1.2)
      nio4r (~> 2.0)
      websocket-driver (>= 0.6.1)
      zeitwerk (~> 2.6)
    actionmailbox (7.1.2)
      actionpack (= 7.1.2)
      activejob (= 7.1.2)
      activerecord (= 7.1.2)
      activestorage (= 7.1.2)
      activesupport (= 7.1.2)
      mail (>= 2.7.1)
      net-imap
      net-pop
      net-smtp
    actionmailer (7.1.2)
      actionpack (= 7.1.2)
      actionview (= 7.1.2)
      activejob (= 7.1.2)
      activesupport (= 7.1.2)
      mail (~> 2.5, >= 2.5.4)
      net-imap
      net-pop
      net-smtp
      rails-dom-testing (~> 2.2)
    actionpack (7.1.2)
      actionview (= 7.1.2)
      activesupport (= 7.1.2)
      nokogiri (>= 1.8.5)
      racc
      rack (>= 2.2.4)
      rack-session (>= 1.0.1)
      rack-test (>= 0.6.3)
      rails-dom-testing (~> 2.2)
      rails-html-sanitizer (~> 1.6)
    actiontext (7.1.2)
      actionpack (= 7.1.2)
      activerecord (= 7.1.2)
      activestorage (= 7.1.2)
      activesupport (= 7.1.2)
      globalid (>= 0.6.0)
      nokogiri (>= 1.8.5)
    actionview (7.1.2)
      activesupport (= 7.1.2)
      builder (~> 3.1)
      erubi (~> 1.11)
      rails-dom-testing (~> 2.2)
      rails-html-sanitizer (~> 1.6)
    activejob (7.1.2)
      activesupport (= 7.1.2)
      globalid (>= 0.3.6)
    activemodel (7.1.2)
      activesupport (= 7.1.2)
    activerecord (7.1.2)
      activemodel (= 7.1.2)
      activesupport (= 7.1.2)
      timeout (>= 0.4.0)
    activestorage (7.1.2)
      actionpack (= 7.1.2)
      activejob (= 7.1.2)
      activerecord (= 7.1.2)
      activesupport (= 7.1.2)
      marcel (~> 1.0)
    activesupport (7.1.2)
      base64
      bigdecimal
      concurrent-ruby (~> 1.0, >= 1.0.2)
      connection_pool (>= 2.2.5)
      drb
      i18n (>= 1.6, < 2)
      minitest (>= 5.1)
      mutex_m
      tzinfo (~> 2.0)
    addressable (2.8.6)
      public_suffix (>= 2.0.2, < 6.0)
    base64 (0.2.0)
    bigdecimal (3.1.5)
    builder (3.2.4)
    carrierwave (3.0.5)
      activemodel (>= 6.0.0)
      activesupport (>= 6.0.0)
      addressable (~> 2.6)
      image_processing (~> 1.1)
      marcel (~> 1.0.0)
      ssrf_filter (~> 1.0)
    combustion (1.3.7)
      activesupport (>= 3.0.0)
      railties (>= 3.0.0)
      thor (>= 0.14.6)
    concurrent-ruby (1.2.2)
    connection_pool (2.4.1)
    content_disposition (1.0.0)
    crass (1.0.6)
    date (3.3.4)
    down (5.4.1)
      addressable (~> 2.8)
    drb (2.2.0)
      ruby2_keywords
    erubi (1.12.0)
    ffi (1.16.3)
    globalid (1.2.1)
      activesupport (>= 6.1)
    i18n (1.14.1)
      concurrent-ruby (~> 1.0)
    image_processing (1.12.2)
      mini_magick (>= 4.9.5, < 5)
      ruby-vips (>= 2.0.17, < 3)
    io-console (0.7.1)
    irb (1.11.0)
      rdoc
      reline (>= 0.3.8)
    loofah (2.22.0)
      crass (~> 1.0.2)
      nokogiri (>= 1.12.0)
    mail (2.8.1)
      mini_mime (>= 0.1.1)
      net-imap
      net-pop
      net-smtp
    marcel (1.0.2)
    mini_magick (4.12.0)
    mini_mime (1.1.5)
    minitest (5.20.0)
    mutex_m (0.2.0)
    mysql2 (0.5.5)
    net-imap (0.4.9)
      date
      net-protocol
    net-pop (0.1.2)
      net-protocol
    net-protocol (0.2.2)
      timeout
    net-smtp (0.4.0)
      net-protocol
    nio4r (2.7.0)
    nokogiri (1.15.5-arm64-darwin)
      racc (~> 1.4)
    pg (1.5.4)
    psych (5.1.2)
      stringio
    public_suffix (5.0.4)
    racc (1.7.3)
    rack (3.0.8)
    rack-session (2.0.0)
      rack (>= 3.0.0)
    rack-test (2.1.0)
      rack (>= 1.3)
    rackup (2.1.0)
      rack (>= 3)
      webrick (~> 1.8)
    rails (7.1.2)
      actioncable (= 7.1.2)
      actionmailbox (= 7.1.2)
      actionmailer (= 7.1.2)
      actionpack (= 7.1.2)
      actiontext (= 7.1.2)
      actionview (= 7.1.2)
      activejob (= 7.1.2)
      activemodel (= 7.1.2)
      activerecord (= 7.1.2)
      activestorage (= 7.1.2)
      activesupport (= 7.1.2)
      bundler (>= 1.15.0)
      railties (= 7.1.2)
    rails-dom-testing (2.2.0)
      activesupport (>= 5.0.0)
      minitest
      nokogiri (>= 1.6)
    rails-html-sanitizer (1.6.0)
      loofah (~> 2.21)
      nokogiri (~> 1.14)
    railties (7.1.2)
      actionpack (= 7.1.2)
      activesupport (= 7.1.2)
      irb
      rackup (>= 1.0.0)
      rake (>= 12.2)
      thor (~> 1.0, >= 1.2.2)
      zeitwerk (~> 2.6)
    rake (13.1.0)
    rbnacl (7.1.1)
      ffi
    rdoc (6.6.2)
      psych (>= 4.0.0)
    reline (0.4.1)
      io-console (~> 0.5)
    ruby-vips (2.2.0)
      ffi (~> 1.12)
    ruby2_keywords (0.0.5)
    shrine (3.5.0)
      content_disposition (~> 1.0)
      down (~> 5.1)
    sqlite3 (1.6.9-arm64-darwin)
    ssrf_filter (1.1.2)
    stringio (3.1.0)
    thor (1.3.0)
    timeout (0.4.1)
    tzinfo (2.0.6)
      concurrent-ruby (~> 1.0)
    webrick (1.8.1)
    websocket-driver (0.7.6)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.5)
    zeitwerk (2.6.12)

PLATFORMS
  arm64-darwin-22

DEPENDENCIES
  carrierwave (~> 3)
  combustion (>= 1.3)
  lockbox!
  minitest (>= 5)
  mysql2
  pg
  rails (~> 7.1.0)
  rake
  rbnacl (>= 6)
  shrine
  sqlite3

BUNDLED WITH
   2.4.10
```

**Comment by ankane** [comment]
Hey @kaoru, pushed a fix in the commit above. Let me know if that works for you, and I'll push a new release.

**Comment by kaoru** [comment]
Hey @ankane, I'm happy to report that that fix seems to work well for our use cases 😄 Thanks so much for investigating further and finding a better fix, and for lockbox in general! 🙏

**Comment by ankane** [comment]
Great, I'm not exactly sure it's better, but seems less likely to break something else if it's more limited. Just released 1.3.1. Thanks again!
